### ISSUE TITLE
Define and implement a caching strategy

### ISSUE THREAD

**Issue by beeme1mr** [open-feature/flagd#216]
## Overview

Flag evaluation using FlagD is done via inter-process communication. While the overhead is minimal, it can be effectively eliminated with an in-memory cache.

## Proposal

Implement a basic caching strategy in FlagD providers that support server-to-client streaming. Flags without a targeting rule defined can cache flags flag until the configuration changes. The server-to-client stream can be used to bust the cache when a flag is updated.

## Considerations

- Cache busting requires a change event.
- Connection errors with FlagD should bust the cache

## Requirements

- [x] #223
- [x] Update FlagD to return caching metadata
- [x] Update at least one FlagD provider to include a cache - https://github.com/open-feature/go-sdk-contrib/issues/44

**Comment by skyerus** [comment]
@beeme1mr 

Is a doc in the flagd repo the best place to document the strategy? We can give guidelines there but ultimately it's up to the providers to implement it.

What caching metadata did you have in mind? The `Reason` is already an emitted field by flagd, I think it's sufficient to cache the value based on  `Reason != "TARGETING_MATCH"`, if you agree I don't think there's any need for flagd to emit anymore data at this stage.

**Comment by toddbaert** [comment]
`TARGETING_MATCH` (and reasons in general) provide some helpful data on caching, but they don't solve the entire problem. One of the most valuable things to cache would be flag values with no rules, but the reason for any particular evaluation doesn't tell us that rules don't exist. For example, a set of targeting rules that don't match at all will default, and so the `DEFAULT` reason is returned. That doesn't mean that we can cache that value, since the next evaluation may have different context and match a rule.

We may be able to solve this only with additional reason codes and changes to reason semantics within flagd, but I think some additional work is needed.

**Comment by skyerus** [comment]
@toddbaert good point, thanks, I hadn't considered that flagd would return `DEFAULT` if there are rules that simply weren't met.

My preference would be to use the [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) HTTP header for this, we can set it to `no-store` in the cases where dynamic evaluation is applicable. Otherwise leave it empty, I don't think setting a `max-age` is necessary given that it ought to survive until busted.

**Comment by AlexsJones** [comment]
I would like to have caching optional personally. I have a few scenarios where I am using `flagD` in the kernel and it's not something I can see a use for if I am interacting via curl - Though in that situation, looking at the above, are you suggesting we set a cache-control header at 0 to avoid caching?

**Comment by skyerus** [comment]
@AlexsJones this documents the current strategy: https://github.com/open-feature/flagd/pull/224/files#diff-68a93cfd67d08237c7e69ffd89a7438adeb9b3671cb292d07803e0cdba4daba3
Feedback, as always, is welcome

TLDR; it's up to clients to decide whether to cache or not, flags that are safe to cache have the reason `STATIC` (we've decided against using the cache-control header). In the future this may be extended to handle dynamic evaluations too.

**Comment by skyerus** [comment]
Closed with https://github.com/open-feature/go-sdk-contrib/issues/44
